File: converted_source/uscommands.pas
unit uscommands;

interface

uses Windows, Graphics, Classes, Controls, SysUtils, StdCtrls, UCommand, USWorld, 
		USFocusCommands, USVariableCommands;

type

{TEditChangeCommand = class(KfCommand)
  	entryField: TObject;
    oldSelectionInformation: TSelectionInformation;
    newSelectionInformation: TSelectionInformation;
		closed: boolean;
		procedure recordChangesMadeToEdit;
		constructor createWithEdit(Sender: TObject;  const lastSelectionInformation: TSelectionInformation);
 		procedure doCommand; override;
		procedure undoCommand; override;
		procedure redoCommand; override;
		function description: string; override;
		end;

  	entryField: TObject;
    newSelectionInformation: TSelectionInformation;
		constructor createWithRule(Sender: TObject;  const lastSelectionInformation: TSelectionInformation);
   }

	TSRuleFieldChange = class(KfCommand)
    rule: TSRule;
    field: integer;
    oldValue: string;
    newValue: string;
		constructor createWithRule(rule: TSRule; field: integer; const newValue: string);
    procedure updateEditorForChange;
 		procedure doCommand; override;
		procedure undoCommand; override;
		procedure redoCommand; override;
		function description: string; override;
    end;

	TSNewRulesCommand = class(KfCommand)
    rules: TList;
    creator: string;
    constructor create;
    procedure addRule(rule: TSRule);
    destructor destroy; override;
  	procedure doCommand; override;
  	procedure undoCommand; override;
  	procedure redoCommand; override;
  	function description: string; override;
    end;

	TSDeleteRulesCommand = class(KfCommand)
    ruleWrappers: TList;
    constructor create;
    procedure addRule(rule: TSRule; newIndex: integer);
    destructor destroy; override;
  	procedure doCommand; override;
  	procedure undoCommand; override;
  	procedure redoCommand; override;
  	function description: string; override;
    end;

	TSMoveRulesCommand = class(KfCommand)
    ruleWrappers: TList;
    action: string;
    constructor create;
    procedure addRule(rule: TSRule; newIndex: integer);
    destructor destroy; override;
  	procedure doCommand; override;
  	procedure undoCommand; override;
  	procedure redoCommand; override;
  	function description: string; override;
    end;

	TSMapDragCommand = class(KfCommand)
    dragRecords: TList;
  	notifyProcedure: TCommandEvent;
    constructor create;
    destructor destroy; override;
  	procedure doCommand; override;
  	procedure undoCommand; override;
  	procedure redoCommand; override;
  	function description: string; override;
  	function TrackMouse(aTrackPhase: TrackPhase; var anchorPoint, previousPoint, nextPoint: TPoint;
  		mouseDidMove, rightButtonDown: boolean): KfCommand; override;
    end;

	TSCommandList = class(KfCommandList)
    function toggleVariable(variable: TSVariable): TSToggleVariableCommand;
    function moveFocus(newFocus: TSVariable): TSMoveFocusCommand;
		function doCommandPhrase(const commandPhrase: string): TSDoCommandPhrase;
		function ruleFieldChange(rule: TSRule; field: integer; const newValue: string): TSRuleFieldChange;
    end;

implementation

uses USDomain, USConsoleForm, USSpeech, USRuleEditorForm, USChangeLog;

{ TSRuleFieldChange ------------------------------------------}
constructor TSRuleFieldChange.createWithRule(rule: TSRule; field: integer; const newValue: string);
  begin
  create;
  self.rule := rule;
  self.field := field;
  self.oldValue := rule.getTextForField(field);
  self.newValue := newValue;
  end;

procedure TSRuleFieldChange.updateEditorForChange;
  begin
  RuleEditorForm.rule := rule;
  RuleEditorForm.loadAllRuleFields;
  
  if (field = kRuleContext) or (field = kRuleMove) then
    ConsoleForm.locationCacheValid := false;
    
 	{wrapper entries will get freed if list box - so reset them}
 	if field = kRuleRequirements then
  	RuleEditorForm.fillListBox(RuleEditorForm.requirementsListBox, rule.requirements)
  else if field = kRuleChanges then
		RuleEditorForm.fillListBox(RuleEditorForm.changesListBox, rule.changes);

 	RuleEditorForm.RuleGrid.invalidate;
 	RuleEditorForm.MapPaintBoxChanged;
 	RuleEditorForm.SecondListBox.invalidate; // could optimize to only do in certain cases
  if RuleEditorForm.organizeByField = field then   // could optimize to only do when browser visible
    RuleEditorForm.setOrganizeByField(RuleEditorForm.organizeByField);
  end;

procedure TSRuleFieldChange.doCommand;
  begin
  Domain.world.lastVariableCreated := '';
  rule.setTextForField(field, newValue);
  // log changes
  if field <> kRuleReply then
  	ChangeLogForm.addToLog(Domain.world.lastVariableCreated)
  else
    ChangeLogForm.addToLog(newValue);
  self.updateEditorForChange;
  inherited doCommand;
  end;

procedure TSRuleFieldChange.undoCommand;
  begin
  rule.setTextForField(field, oldValue);
  self.updateEditorForChange;
  RuleEditorForm.selectEditorField(field);
  inherited undoCommand;
  end;

procedure TSRuleFieldChange.redoCommand;
  begin
  rule.setTextForField(field, newValue);
  self.updateEditorForChange;
  RuleEditorForm.selectEditorField(field);
  inherited doCommand;
  end;

function TSRuleFieldChange.description: string;
  begin
//  result := 'rule ' + IntToStr(domain.world.rules.indexOf(rule) + 1) + ' change of ' + TSRule.headerForField(field);
  result := 'Change ' + TSRule.headerForField(field) + ' For Rule ' + IntToStr(domain.world.rules.indexOf(rule) + 1);
  end;


{ TSNewRulesCommand ------------------------------------------}

constructor TSNewRulesCommand.create;
  begin
  rules := TList.create;
  end;

destructor TSNewRulesCommand.destroy;
  var
  	i: integer;
  begin
  if not self.done then
  	if rules <> nil then
    	for i := 0 to rules.count - 1 do
      	TObject(rules[i]).free;
  rules.free;
  rules := nil;
  end;

procedure TSNewRulesCommand.addRule(rule: TSRule);
  begin
  rules.add(rule);
  end;

procedure TSNewRulesCommand.doCommand;
  begin
  {already added at start}
	inherited doCommand;
  RuleEditorForm.updateForRuleChange;
  RuleEditorForm.scrollGridSelectionsIntoView(kFromBottom);
  end;

procedure TSNewRulesCommand.undoCommand;
  var
  	i: integer;
    rule: TSRule;
  begin
  for i := 0 to rules.count - 1 do
    begin
    rule := TSRule(rules[i]);
    domain.world.rules.remove(rule);
    rule.selected := false;
    rule.removeUseages;
    end;
  inherited undoCommand;
  if rules.IndexOf(RuleEditorForm.rule) >= 0 then
    RuleEditorForm.editRule(nil); 
  RuleEditorForm.updateForRuleChange;
  end;

procedure TSNewRulesCommand.redoCommand;
  var
  	i: integer;
    rule: TSRule;
  begin
  domain.world.deselectAllExcept(nil);
  for i := 0 to rules.count - 1 do
    begin
    rule := TSRule(rules[i]);
    rule.selected := true;
    domain.world.rules.add(rule);
    rule.addUseages;
    end;
	inherited doCommand;
  RuleEditorForm.updateForRuleChange;
  RuleEditorForm.scrollGridSelectionsIntoView(kFromBottom);
  //if rules.count > 0 then
  //  RuleEditorForm.editRule(rules[rules.count - 1]);
  end;

function TSNewRulesCommand.description: string;
  begin
  if rules.count > 1 then
    result := 'new rules'
  else if rules.count = 1 then
    result := 'new rule'
  else
    result := 'new rule';
  if creator <> '' then
    result := result + ' from ' + creator;
  if creator = 'duplicating' then
    result := 'duplicate rule';
  end;

{ TSDeleteRulesCommand ------------------------------------------}

constructor TSDeleteRulesCommand.create;
  begin
  ruleWrappers := TList.create;
  end;

destructor TSDeleteRulesCommand.destroy;
  var
  	i: integer;
    rule: TSRule;
    wrapper: TSIndexChangeRuleWrapper;
  begin
  if ruleWrappers <> nil then
    for i := 0 to ruleWrappers.count - 1 do
      begin
      wrapper := TSIndexChangeRuleWrapper(ruleWrappers[i]);
      rule := wrapper.rule;
      if self.done then
        rule.free;
      wrapper.free;
      end;
  ruleWrappers.free;
  ruleWrappers := nil;
  end;

procedure TSDeleteRulesCommand.addRule(rule: TSRule; newIndex: integer);
  var
    wrapper: TSIndexChangeRuleWrapper;
  begin
  wrapper := TSIndexChangeRuleWrapper.createWithRuleNewIndex(rule, newIndex);
  ruleWrappers.add(wrapper);
  end;

procedure TSDeleteRulesCommand.doCommand;
  var
  	i: integer;
    wrapper: TSIndexChangeRuleWrapper;
  begin
  for i := ruleWrappers.count - 1 downto 0 do
    begin
    wrapper := TSIndexChangeRuleWrapper(ruleWrappers[i]);
    if (wrapper.rule = RuleEditorForm.rule) then
    	RuleEditorForm.editRule(nil);
    wrapper.rule.removeUseages;
    wrapper.doChange;
    end;
	inherited doCommand;
  RuleEditorForm.updateForRuleChange;
  end;

procedure TSDeleteRulesCommand.undoCommand;
  var
  	i: integer;
    wrapper: TSIndexChangeRuleWrapper;
  begin
  domain.world.deselectAllExcept(nil);
  for i := 0 to ruleWrappers.count - 1 do
    begin
    wrapper := TSIndexChangeRuleWrapper(ruleWrappers[i]);
    wrapper.rule.addUseages;
    wrapper.undoChange;
    wrapper.rule.selected := true;
    end;
  if ruleWrappers.count > 0 then
    	RuleEditorForm.editRule(TSIndexChangeRuleWrapper(ruleWrappers[0]).rule);
  inherited undoCommand;
  RuleEditorForm.updateForRuleChange;
  RuleEditorForm.scrollGridSelectionsIntoView(kFromTop)
  end;

procedure TSDeleteRulesCommand.redoCommand;
  var
  	i: integer;
    wrapper: TSIndexChangeRuleWrapper;
  begin
  domain.world.deselectAllExcept(nil);
  for i := ruleWrappers.count - 1 downto 0 do
    begin
    wrapper := TSIndexChangeRuleWrapper(ruleWrappers[i]);
    if (wrapper.rule = RuleEditorForm.rule) then
    	RuleEditorForm.editRule(nil);
    wrapper.rule.removeUseages;
    wrapper.doChange;
    end;
	inherited doCommand;
  RuleEditorForm.updateForRuleChange;
  end;

function TSDeleteRulesCommand.description: string;
  begin
  if ruleWrappers.count > 1 then
    result := 'delete rules'
  else if ruleWrappers.count = 1 then
    result := 'delete rule'
  else
    result := 'delete rule';
  end;

{ TSMoveRulesCommand ------------------------------------------}

constructor TSMoveRulesCommand.create;
  begin
  ruleWrappers := TList.create;
  end;

destructor TSMoveRulesCommand.destroy;
  var
  	i: integer;
    wrapper: TSIndexChangeRuleWrapper;
  begin
  if ruleWrappers <> nil then
    for i := 0 to ruleWrappers.count - 1 do
      begin
      wrapper := TSIndexChangeRuleWrapper(ruleWrappers[i]);
      wrapper.free;
      end;
  ruleWrappers.free;
  ruleWrappers := nil;
  end;

procedure TSMoveRulesCommand.addRule(rule: TSRule; newIndex: integer);
  var
    wrapper: TSIndexChangeRuleWrapper;
  begin
  wrapper := TSIndexChangeRuleWrapper.createWithRuleNewIndex(rule, newIndex);
  ruleWrappers.add(wrapper);
  end;

procedure TSMoveRulesCommand.doCommand;
  var
  	i: integer;
    wrapper: TSIndexChangeRuleWrapper;
  begin
  for i := 0 to ruleWrappers.count - 1 do
    begin
    wrapper := TSIndexChangeRuleWrapper(ruleWrappers[i]);
    wrapper.doChange;
    end;
	inherited doCommand;
  RuleEditorForm.RuleGrid.invalidate;
  if action = 'raise' then
  	RuleEditorForm.scrollGridSelectionsIntoView(kFromTop)
  else
  	RuleEditorForm.scrollGridSelectionsIntoView(kFromBottom);
  RuleEditorForm.updateRuleNumberLabel;
  end;

procedure TSMoveRulesCommand.undoCommand;
  var
  	i: integer;
    wrapper: TSIndexChangeRuleWrapper;
  begin
  domain.world.deselectAllExcept(nil);
  for i := ruleWrappers.count - 1 downto 0 do
    begin
    wrapper := TSIndexChangeRuleWrapper(ruleWrappers[i]);
    wrapper.rule.selected := true;
    wrapper.undoChange;
    end;
  inherited undoCommand;
  RuleEditorForm.RuleGrid.invalidate;
  if action = 'raise' then
  	RuleEditorForm.scrollGridSelectionsIntoView(kFromBottom)
  else
  	RuleEditorForm.scrollGridSelectionsIntoView(kFromTop);
  RuleEditorForm.updateRuleNumberLabel;
  end;

procedure TSMoveRulesCommand.redoCommand;
  var
  	i: integer;
    wrapper: TSIndexChangeRuleWrapper;
  begin
  domain.world.deselectAllExcept(nil);
  for i := 0 to ruleWrappers.count - 1 do
    begin
    wrapper := TSIndexChangeRuleWrapper(ruleWrappers[i]);
    wrapper.rule.selected := true;
    wrapper.doChange;
    end;
	inherited doCommand;
  RuleEditorForm.RuleGrid.invalidate;
  if action = 'raise' then
  	RuleEditorForm.scrollGridSelectionsIntoView(kFromTop)
  else
  	RuleEditorForm.scrollGridSelectionsIntoView(kFromBottom);
  RuleEditorForm.updateRuleNumberLabel;
  end;

function TSMoveRulesCommand.description: string;
  begin
  if ruleWrappers.count > 1 then
    result := 'rules'
  else if ruleWrappers.count = 1 then
    result := 'rule'
  else
    result := 'rule';
  if action <> '' then
    result := action + ' ' + result
  else
    result := 'move ' + result;
  end;

{ TSMapDragCommand ------------------------------------------}

constructor TSMapDragCommand.create;
  begin
  dragRecords := TList.create;
  domain.world.addDragRecordsToList(dragRecords);
  end;

destructor TSMapDragCommand.destroy;
  var
  	i: integer;
  begin
  if dragRecords <> nil then
    for i := 0 to dragRecords.count - 1 do
      TObject(dragRecords[i]).free;
  dragRecords.free;
  dragRecords := nil;
  end;

procedure TSMapDragCommand.doCommand;
  var
  	i: integer;
  begin
  for i := 0 to dragRecords.count - 1 do
    TSDragRecord(dragRecords[i]).doDrag;
	if assigned(notifyProcedure) then
  	notifyProcedure(self, commandDone);
	inherited doCommand;
  end;

procedure TSMapDragCommand.undoCommand;
  var
  	i: integer;
  begin
  domain.world.deselectAllExcept(nil);
  for i := 0 to dragRecords.count - 1 do
    begin
    TSDragRecord(dragRecords[i]).draggedNode.selected := true;
    TSDragRecord(dragRecords[i]).undoDrag;
    end;
	if assigned(notifyProcedure) then
  	notifyProcedure(self, commandUndone);
  inherited undoCommand;
  end;

procedure TSMapDragCommand.redoCommand;
  var
  	i: integer;
  begin
  domain.world.deselectAllExcept(nil);
  for i := 0 to dragRecords.count - 1 do
    begin
    TSDragRecord(dragRecords[i]).draggedNode.selected := true;
    TSDragRecord(dragRecords[i]).doDrag;
    end;
	if assigned(notifyProcedure) then
  	notifyProcedure(self, commandDone);
	inherited doCommand;
  end;

function TSMapDragCommand.description: string;
  begin
  if dragRecords.count > 1 then
    result := 'Drag nodes'
  else if dragRecords.count = 1 then
    result := 'Drag ' + TSDragRecord(dragRecords[0]).draggedNode.displayName
  else
    result := 'Drag';
  end;

function TSMapDragCommand.TrackMouse(aTrackPhase: TrackPhase; var anchorPoint, previousPoint, nextPoint: TPoint;
		mouseDidMove, rightButtonDown: boolean): KfCommand;
  var
  	i: integer;
    delta: TPoint;
  begin
  result := self;
  case aTrackPhase of
  	trackPress:
      begin
      if dragRecords.count = 0 then
        begin
        result := nil;
        self.free;
        exit;
        end;
      end;
    trackMove:
      begin
      if mouseDidMove then
        begin
      	delta :=
      		Point(nextPoint.x - previousPoint.x,
      					nextPoint.y - previousPoint.y);
  			for i := 0 to dragRecords.count - 1 do
    			TSDragRecord(dragRecords[i]).offset(delta);
				if assigned(notifyProcedure) then
  				notifyProcedure(self, commandDone);
        end;
      end;
  	trackRelease:
  		begin
     	if not mouseDidMove then
      	begin
   			if (TSDragRecord(dragRecords[0]).draggedNode.position.x <>
        			TSDragRecord(dragRecords[0]).originalLocation.x) or
        	 (TSDragRecord(dragRecords[0]).draggedNode.position.y <>
           		TSDragRecord(dragRecords[0]).originalLocation.y) then
          begin
  				for i := 0 to dragRecords.count - 1 do
    				TSDragRecord(dragRecords[i]).undoDrag;
					if assigned(notifyProcedure) then
  					notifyProcedure(self, commandDone);
          end;
      	result := nil;
      	self.free;
      	end
      else
        begin
       	delta :=
      		Point(nextPoint.x - previousPoint.x,
      					nextPoint.y - previousPoint.y);
   			if (delta.x <> 0) or (delta.y <> 0) then
          begin
  			  for i := 0 to dragRecords.count - 1 do
    			  TSDragRecord(dragRecords[i]).offset(delta);
					if assigned(notifyProcedure) then
  					notifyProcedure(self, commandDone);
          end;
        end;
  		end;
    end;
  end;

{ ----------------------------- TSCommandList -------------------------------}

function TSCommandList.toggleVariable(variable: TSVariable): TSToggleVariableCommand;
  begin
  result := TSToggleVariableCommand.createWithVariable(variable);
  self.doCommand(result);
  end;

function TSCommandList.moveFocus(newFocus: TSVariable): TSMoveFocusCommand;
  begin
  result := TSMoveFocusCommand.createWithNewFocus(newFocus);
  self.doCommand(result);
  end;

function TSCommandList.doCommandPhrase(const commandPhrase: string): TSDoCommandPhrase;
  begin
  result := TSDoCommandPhrase.createWithCommandPhrase(commandPhrase);
  self.doCommand(result);
  end;

function TSCommandList.ruleFieldChange(rule: TSRule; field: integer; const newValue: string): TSRuleFieldChange;
  var
    newContextOrMove: TSVariable;
  begin
  if (field = kRuleContext) or (field = kRuleMove) then
    if Pos('new context ', rule.getTextForField(field)) = 1 then
      begin
      if domain.world.findVariable(newValue) = nil then
        begin
   			newContextOrMove := domain.world.findOrCreateVariable(newValue, false);
    		newContextOrMove.position := rule.context.position;
        end;
      end;
  result := TSRuleFieldChange.createWithRule(rule, field, newValue);
  self.doCommand(result);
  end;

end.

File: source/uscommands.ts
// unit uscommands

import { int, arrayRemove} from "./common"
import { KfCommand, TrackPhase, KfCommandChangeType, TCommandEvent } from "./KfCommand"
import { TSRule, TSRuleField } from "./TSRule"
import { TSIndexChangeRuleWrapper } from "./TSIndexChangeRuleWrapper"
import { TPoint } from "./TPoint"
import { TSDragRecord } from "./TSDragRecord"
import { TWorld } from "./TWorld";

// TODO: Fix these as imports
import { RuleEditorForm, ChangeLogForm, ConsoleForm, ScrollIntoViewDirection } from "./fixTypes"

export class TSRuleFieldChangeCommand extends KfCommand {
    world: TWorld
    ruleEditorForm: RuleEditorForm
    changeLogForm: ChangeLogForm
    consoleForm: ConsoleForm
    rule: TSRule
    field: int
    oldValue: string
    newValue: string

    constructor(world: TWorld, ruleEditorForm: RuleEditorForm, changeLogForm: ChangeLogForm, consoleForm: ConsoleForm, rule: TSRule, field: int, newValue: string) {
        super()
        this.world = world
        this.ruleEditorForm = ruleEditorForm
        this.changeLogForm = changeLogForm
        this.consoleForm = consoleForm
        this.rule = rule
        this.field = field
        this.oldValue = rule.getTextForField(field)
        this.newValue = newValue
    }
    
    updateEditorForChange(): void {
        this.ruleEditorForm.rule = this.rule
        this.ruleEditorForm.loadAllRuleFields()
        // TODO: This locationCacheValid field probably can be removed along with consoleForm after finish refactoring
        if ((this.field === TSRuleField.kRuleContext) || (this.field === TSRuleField.kRuleMove)) {
            this.consoleForm.locationCacheValid = false
        }
        if (this.field === TSRuleField.kRuleRequirements) {
            //wrapper entries will get freed if list box - so reset them
            this.ruleEditorForm.fillListBox(this.ruleEditorForm.RequirementsListBox, this.rule.requirements)
        } else if (this.field === TSRuleField.kRuleChanges) {
            this.ruleEditorForm.fillListBox(this.ruleEditorForm.ChangesListBox, this.rule.changes)
        }
        this.ruleEditorForm.RuleGrid.Invalidate()
        this.ruleEditorForm.MapPaintBoxChanged()
        // could optimize to only do in certain cases
        this.ruleEditorForm.SecondListBox.Invalidate()
        if (this.ruleEditorForm.organizeByField === this.field) {
            // could optimize to only do when browser visible
            this.ruleEditorForm.setOrganizeByField(this.ruleEditorForm.organizeByField)
        }
    }
    
    doCommand(): void {
        this.world.lastVariableCreated = ""
        this.rule.setTextForField(this.field, this.newValue)
        if (this.field !== TSRuleField.kRuleReply) {
            // log changes
            this.changeLogForm.addToLog(this.world.lastVariableCreated)
        } else {
            this.changeLogForm.addToLog(this.newValue)
        }
        this.updateEditorForChange()
        super.doCommand()
    }
    
    undoCommand(): void {
        this.rule.setTextForField(this.field, this.oldValue)
        this.updateEditorForChange()
        this.ruleEditorForm.selectEditorField(this.field)
        super.undoCommand()
    }
    
    redoCommand(): void {
        this.rule.setTextForField(this.field, this.newValue)
        this.updateEditorForChange()
        this.ruleEditorForm.selectEditorField(this.field)
        super.doCommand()
    }
    
    description(): string {
        let result = ""
        //  result := 'rule ' + IntToStr(domain.world.rules.indexOf(rule) + 1) + ' change of ' + TSRule.headerForField(field);
        result = "Change " + TSRule.headerForField(this.field) + " For Rule " + (this.world.rules.indexOf(this.rule) + 1)
        return result
    }
    
}

export class TSNewRulesCommand extends KfCommand {
    world: TWorld
    ruleEditorForm: RuleEditorForm
    rules: TSRule[] = []
    creator: string = ""

    constructor(world: TWorld, ruleEditorForm: RuleEditorForm) {
        super()
        this.world = world
        this.ruleEditorForm = ruleEditorForm 
    }
    
    addRule(rule: TSRule): void {
        this.rules.push(rule)
    }
    
    doCommand(): void {
        //already added at start
        super.doCommand()
        this.ruleEditorForm.updateForRuleChange()
        this.ruleEditorForm.scrollGridSelectionsIntoView(ScrollIntoViewDirection.kFromBottom)
    }
    
    undoCommand(): void {
        for (let i = 0; i < this.rules.length; i++) {
            const rule: TSRule = this.rules[i]
            arrayRemove(this.world.rules, rule)
            rule.selected = false
            rule.removeUseages()
        }
        super.undoCommand()
        if (this.rules.indexOf(this.ruleEditorForm.rule) >= 0) {
            this.ruleEditorForm.editRule(null)
        }
        this.ruleEditorForm.updateForRuleChange()
    }
    
    redoCommand(): void {
        this.world.deselectAllExcept(null)
        for (let i = 0; i < this.rules.length; i++) {
            const rule: TSRule = this.rules[i]
            rule.selected = true
            this.world.rules.push(rule)
            rule.addUseages()
        }
        super.doCommand()
        this.ruleEditorForm.updateForRuleChange()
        this.ruleEditorForm.scrollGridSelectionsIntoView(ScrollIntoViewDirection.kFromBottom)
        //if rules.count > 0 then
        //  RuleEditorForm.editRule(rules[rules.count - 1]);
    }
    
    description(): string {
        let result = ""
        if (this.rules.length > 1) {
            result = "new rules"
        } else if (this.rules.length === 1) {
            result = "new rule"
        } else {
            result = "new rule"
        }
        if (this.creator !== "") {
            result = result + " from " + this.creator
        }
        if (this.creator === "duplicating") {
            result = "duplicate rule"
        }
        return result
    }
    
}

export class TSDeleteRulesCommand extends KfCommand {
    world: TWorld
    ruleEditorForm: RuleEditorForm
    ruleWrappers: TSIndexChangeRuleWrapper[] = []

    constructor(world: TWorld, ruleEditorForm: RuleEditorForm) {
        super()
        this.world = world
        this.ruleEditorForm = ruleEditorForm
    }
    
    addRule(rule: TSRule, newIndex: int): void {
        const wrapper: TSIndexChangeRuleWrapper = new TSIndexChangeRuleWrapper(rule, newIndex)
        this.ruleWrappers.push(wrapper)
    }
    
    doCommand(): void {
        for (let i = this.ruleWrappers.length - 1; i >= 0; i--) {
            const wrapper: TSIndexChangeRuleWrapper = this.ruleWrappers[i]
            if ((wrapper.rule === this.ruleEditorForm.rule)) {
                this.ruleEditorForm.editRule(null)
            }
            wrapper.rule.removeUseages()
            wrapper.doChange()
        }
        super.doCommand()
        this.ruleEditorForm.updateForRuleChange()
    }
    
    undoCommand(): void {
        this.world.deselectAllExcept(null)
        for (let i = 0; i < this.ruleWrappers.length; i++) {
            const wrapper: TSIndexChangeRuleWrapper = this.ruleWrappers[i]
            wrapper.rule.addUseages()
            wrapper.undoChange()
            wrapper.rule.selected = true
        }
        if (this.ruleWrappers.length > 0) {
            this.ruleEditorForm.editRule(this.ruleWrappers[0].rule)
        }
        super.undoCommand()
        this.ruleEditorForm.updateForRuleChange()
        this.ruleEditorForm.scrollGridSelectionsIntoView(ScrollIntoViewDirection.kFromTop)
    }
    
    redoCommand(): void {
        this.world.deselectAllExcept(null)
        for (let i = this.ruleWrappers.length - 1; i >= 0; i--) {
            const wrapper: TSIndexChangeRuleWrapper = this.ruleWrappers[i]
            if ((wrapper.rule === this.ruleEditorForm.rule)) {
                this.ruleEditorForm.editRule(null)
            }
            wrapper.rule.removeUseages()
            wrapper.doChange()
        }
        super.doCommand()
        this.ruleEditorForm.updateForRuleChange()
    }
    
    description(): string {
        let result = ""
        if (this.ruleWrappers.length > 1) {
            result = "delete rules"
        } else if (this.ruleWrappers.length === 1) {
            result = "delete rule"
        } else {
            result = "delete rule"
        }
        return result
    }
    
}

export class TSMoveRulesCommand extends KfCommand {
    world: TWorld
    ruleEditorForm: RuleEditorForm
    ruleWrappers: TSIndexChangeRuleWrapper[] = []
    action: string = ""

    constructor(world: TWorld, ruleEditorForm: RuleEditorForm) {
        super()
        this.world = world
        this.ruleEditorForm = ruleEditorForm
    }
    
    addRule(rule: TSRule, newIndex: int): void {
        const wrapper: TSIndexChangeRuleWrapper = new TSIndexChangeRuleWrapper(rule, newIndex)
        this.ruleWrappers.push(wrapper)
    }
    
    doCommand(): void {
        for (let i = 0; i < this.ruleWrappers.length; i++) {
            const wrapper: TSIndexChangeRuleWrapper = this.ruleWrappers[i]
            wrapper.doChange()
        }
        super.doCommand()
        this.ruleEditorForm.RuleGrid.Invalidate()
        if (this.action === "raise") {
            this.ruleEditorForm.scrollGridSelectionsIntoView(ScrollIntoViewDirection.kFromTop)
        } else {
            this.ruleEditorForm.scrollGridSelectionsIntoView(ScrollIntoViewDirection.kFromBottom)
        }
        this.ruleEditorForm.updateRuleNumberLabel()
    }
    
    undoCommand(): void {
        this.world.deselectAllExcept(null)
        for (let i = this.ruleWrappers.length - 1; i >= 0; i--) {
            const wrapper: TSIndexChangeRuleWrapper = this.ruleWrappers[i]
            wrapper.rule.selected = true
            wrapper.undoChange()
        }
        super.undoCommand()
        this.ruleEditorForm.RuleGrid.Invalidate()
        if (this.action === "raise") {
            this.ruleEditorForm.scrollGridSelectionsIntoView(ScrollIntoViewDirection.kFromBottom)
        } else {
            this.ruleEditorForm.scrollGridSelectionsIntoView(ScrollIntoViewDirection.kFromTop)
        }
        this.ruleEditorForm.updateRuleNumberLabel()
    }
    
    redoCommand(): void {
        this.world.deselectAllExcept(null)
        for (let i = 0; i < this.ruleWrappers.length; i++) {
            const wrapper: TSIndexChangeRuleWrapper = this.ruleWrappers[i]
            wrapper.rule.selected = true
            wrapper.doChange()
        }
        super.doCommand()
        this.ruleEditorForm.RuleGrid.Invalidate()
        if (this.action === "raise") {
            this.ruleEditorForm.scrollGridSelectionsIntoView(ScrollIntoViewDirection.kFromTop)
        } else {
            this.ruleEditorForm.scrollGridSelectionsIntoView(ScrollIntoViewDirection.kFromBottom)
        }
        this.ruleEditorForm.updateRuleNumberLabel()
    }
    
    description(): string {
        let result = ""
        if (this.ruleWrappers.length > 1) {
            result = "rules"
        } else if (this.ruleWrappers.length === 1) {
            result = "rule"
        } else {
            result = "rule"
        }
        if (this.action !== "") {
            result = this.action + " " + result
        } else {
            result = "move " + result
        }
        return result
    }
    
}

export class TSMapDragCommand extends KfCommand {
    world: TWorld
    dragRecords: TSDragRecord[] = []
    notifyProcedure: TCommandEvent
    
    constructor(world: TWorld) {
        super()
        this.world = world
        this.world.addDragRecordsToList(this.dragRecords)
    }
    
    doCommand(): void {
        for (let i = 0; i < this.dragRecords.length; i++) {
            this.dragRecords[i].doDrag()
        }
        if (this.notifyProcedure) {
            this.notifyProcedure(this, KfCommandChangeType.commandDone)
        }
        super.doCommand()
    }
    
    undoCommand(): void {
        this.world.deselectAllExcept(null)
        for (let i = 0; i < this.dragRecords.length; i++) {
            this.dragRecords[i].draggedNode.selected = true
            this.dragRecords[i].undoDrag()
        }
        if (this.notifyProcedure) {
            this.notifyProcedure(this, KfCommandChangeType.commandUndone)
        }
        super.undoCommand()
    }
    
    redoCommand(): void {
        this.world.deselectAllExcept(null)
        for (let i = 0; i < this.dragRecords.length; i++) {
            this.dragRecords[i].draggedNode.selected = true
            this.dragRecords[i].doDrag()
        }
        if (this.notifyProcedure) {
            this.notifyProcedure(this, KfCommandChangeType.commandDone)
        }
        super.doCommand()
    }
    
    description(): string {
        let result = ""
        if (this.dragRecords.length > 1) {
            result = "Drag nodes"
        } else if (this.dragRecords.length === 1) {
            result = "Drag " + this.dragRecords[0].draggedNode.displayName()
        } else {
            result = "Drag"
        }
        return result
    }
    
    trackMouse(aTrackPhase: TrackPhase, anchorPoint: TPoint, previousPoint: TPoint, nextPoint: TPoint, mouseDidMove: boolean, rightButtonDown: boolean): KfCommand | null {
        let result: KfCommand | null

        result = this
        switch (aTrackPhase) {
            case TrackPhase.trackPress:
                if (this.dragRecords.length === 0) {
                    result = null
                }
                break
            case TrackPhase.trackMove:
                if (mouseDidMove) {
                    const delta = new TPoint(nextPoint.X - previousPoint.X, nextPoint.Y - previousPoint.Y)
                    for (let i = 0; i < this.dragRecords.length; i++) {
                        this.dragRecords[i].offset(delta)
                    }
                    if (this.notifyProcedure) {
                        this.notifyProcedure(this, KfCommandChangeType.commandDone)
                    }
                }
                break
            case TrackPhase.trackRelease:
                if (!mouseDidMove) {
                    if ((this.dragRecords[0].draggedNode.position.X !== this.dragRecords[0].originalLocation.X) || (this.dragRecords[0].draggedNode.position.Y !== this.dragRecords[0].originalLocation.Y)) {
                        for (let i = 0; i < this.dragRecords.length; i++) {
                            this.dragRecords[i].undoDrag()
                        }
                        if (this.notifyProcedure) {
                            this.notifyProcedure(this, KfCommandChangeType.commandDone)
                        }
                    }
                    result = null
                } else {
                    const delta = new TPoint(nextPoint.X - previousPoint.X, nextPoint.Y - previousPoint.Y)
                    if ((delta.X !== 0) || (delta.Y !== 0)) {
                        for (let i = 0; i < this.dragRecords.length; i++) {
                            this.dragRecords[i].offset(delta)
                        }
                        if (this.notifyProcedure) {
                            this.notifyProcedure(this, KfCommandChangeType.commandDone)
                        }
                    }
                }
                break
        }
        return result
    }
    
}
